src/linter/block.go

```diff
@@ -1080,8 +1080,36 @@ func (b *BlockWalker) handleForeach(s *ir.ForeachStmt) bool {
 
 			b.handleVariableNode(s.Key, meta.TypesMap{}, "foreach_key")
 			if list, ok := s.Variable.(*ir.ListExpr); ok {
-				for _, item := range list.Items {
-					b.handleVariableNode(item.Val, meta.TypesMap{}, "foreach_value")
+				if !meta.IsIndexingComplete() {
```

Comment:
We need to solve this using lazy types. If the result of a foreach is returned in this function and is used in another one, this logic will not allow the type to propagate all the way. The following should also work:

```
<?php
class Foo { public function hello () {} }

function shaper() {
return shape(1, new Foo());
}

function foreacher() {
foreach ([shaper()] as list($_, $foo)) return $foo;
}

function fooer() {
foreacher()->hello(); // <- no errors here
}
```

Reply:
If we add the necessary comments, then this code passes the test.
```php
<?php
class Foo { public function hello () {} }

/**
 * @return shape(Key1:int, Key2:\Foo)
 */
function shaper() {
	return ["Key1" => 1, "Key2" => new Foo()];
}

/**
 * @return \Foo
 */
function foreacher() {
	foreach ([shaper()] as list("Key1" => $_, "Key2" => $foo)) {
		exprtype($foo, "\Foo");
		return $foo;
	}
}

function fooer() {
	$f = foreacher();
	exprtype($f, "\Foo");
	$f->hello(); // <- no errors here
}
```

Reply:
Sure, but noverify doesn't require you to add phpdoc-comments in cases when it can infer types. Requiring annotations to do such things seems impractical to me and in "real" code annotations for simple functions are often omitted so I would prefer that we try to infer as much information from the code as we can so that you don't have to write annotations everywhere just to please the linter.

Reply:
I agree, however, in this code, the `foreacher` function will not always return the `Foo` type, for example, if `shaper` returns an empty array, then the function will return `void`, which the linter correctly complains about if the `foreacher` does not have phpdoc.